### ISSUE TITLE
Improve in-workflow documentation through a new rule `norunparam`: `description` 

### ISSUE THREAD

**Issue by sdruskat** [snakemake/snakemake#2019]
**Is your feature request related to a problem? Please describe.**

To improve reusability/reproducibility, I'd like to be able to document the single rules in my workflows more easily. This can of course be done in the external scripts, or `shell` and `run` directly, but a more accessible, standardized solution would work better IMHO. (Similar to Python docstrings for functions.) As is, `--filegraph` output usability has room for improevement, as rule names aren't always very accessible.

**Describe the solution you'd like**

Add a new `norunparam` called `description` (or similar, e.g. `documentation` or similar) of type `stringliteral` to `rule`. This can be used to document rules, and can be used in, e.g., `--filegraph` output, or reporting (which I haven't used yet).

```python
rule clone_git:
    description: f'Creates a shallow clone of the repository at {GIT_URL}. This repository contains all the dataz.'
    output:
        git = directory(GIT_DIR)
    threads: 
        1
    conda:
        f'{ENVS_DIR}git.yml'
    shell:
        f'git clone --depth 1 {GIT_URL} {{output}}'
```

**Describe alternatives you've considered**

The `message` field could be hacked to do this, but the semantics are off, as it is meant to adapt the runtime output of a workflow, rather than document the single rules.

Alternatively - and I guess that this is already possible - the use of docstrings could be encouraged for this kind of documentation, and docstrings could be reused for reporting and graph outpts, etc. If this is the way forward, it should be documented in the user manual.

```python
rule clone_git:
    '''
    Creates a shallow clone of the repository at {GIT_URL}. This repository contains all the dataz.
    '''
    output:
        git = directory(GIT_DIR)
    threads: 
        1
    conda:
        f'{ENVS_DIR}git.yml'
    shell:
        f'git clone --depth 1 {GIT_URL} {{output}}'
```

**Comment by pvandyken** [comment]
Indeed, this is the purpose for the docstring... it's available within the api under `Rule.docstring`. AFAIK, currently it's only used with `snakemake --list`, which lists all the rules with their docstrings and other relevant info. I'm sure it could find increased utility within reports (although like you, I haven't used reports much), but I don't think I'd want it included in any of the dag graphs, as it would make them too busy.

**Comment by sdruskat** [comment]
> Indeed, this is the purpose for the docstring... it's available within the api under `Rule.docstring`. AFAIK, currently it's only used with `snakemake --list`, which lists all the rules with their docstrings and other relevant info. I'm sure it could find increased utility within reports (although like you, I haven't used reports much), but I don't think I'd want it included in any of the dag graphs, as it would make them too busy.

Thanks, happy New Year, and please excuse the delay caused by holidays. Thanks for the info about reusing docstrings via `--list`.

As for the DAGs supporting docstring rendering, I think docstrings could be reused here optionally (via flag) perhaps? Depending on rule naming, docstring length, etc., this could become helpful for some scenarios, although I agree that it shouldn't be the norm to include them in DAG rendering.

**Comment by sdruskat** [comment]
> If you wanted to contribute a PR to this effect, I'm sure the work would be welcome.

Thanks. I'd love to do this, but need to find the time.
